### PR TITLE
feat: AWS v2: Split users and credential report users into separate resources

### PR DESCRIPTION
**Pull request by hermanschaaf**

### CODE REVIEW

File: plugins/source/aws/resources/services/iam/users_fetch.go [hunk 1]
```diff
@@ -2,116 +2,67 @@ package iam
...
-			Cert1LastRotated:      reportUser.Cert1LastRotated,
-			Cert2LastRotated:      reportUser.Cert2LastRotated,
-			reportUser:            reportUser,
-			isRoot:                true,
+	listUser := listInfo.(types.User)
+	arn := ""
+	if listUser.Arn != nil {
+		arn = *listUser.Arn
+	}
+	if arn == c.AccountGlobalARN(client.IamService, "root") {
```
Comment (bbernays) [inline plugins/source/aws/resources/services/iam/users_fetch.go:50]: I don't think this check is needed anymore.

File: plugins/source/aws/resources/services/iam/users_fetch.go [hunk 2]
```diff
@@ -2,116 +2,67 @@ package iam
...
+		// TODO: check if this is still necessary after unbundling credential reports
```
Comment (bbernays) [inline plugins/source/aws/resources/services/iam/users_fetch.go:51]: I don't think it is. We should be able to to just return the User object

File: plugins/source/aws/resources/services/iam/users_fetch.go (hunk 1 above)
Comment (bbernays) [inline plugins/source/aws/resources/services/iam/users_fetch.go:50]: I don't think any of this is necessary anymore as `ListUser` never returns the root user

File: plugins/source/aws/resources/services/iam/users_fetch.go [hunk 3]
```diff
@@ -354,7 +163,7 @@ func postIamUserAccessKeyResolver(ctx context.Context, meta schema.ClientMeta, r
 
 func fetchIamUserAttachedPolicies(ctx context.Context, meta schema.ClientMeta, parent *schema.Resource, res chan<- interface{}) error {
 	var config iam.ListAttachedUserPoliciesInput
-	p := parent.Item.(UserWrapper)
+	p := parent.Item.(*types.User)
 	if aws.ToString(p.UserName) == rootName {
 		return nil
 	}
```
Comment (bbernays) [inline plugins/source/aws/resources/services/iam/users_fetch.go:169]: Not necessary anymore as Root user is never returned

File: plugins/source/aws/resources/services/iam/users_fetch.go [hunk 4]
```diff
@@ -289,7 +98,7 @@ func fetchIamUserGroups(ctx context.Context, meta schema.ClientMeta, parent *sch
 
 func fetchIamUserAccessKeys(ctx context.Context, meta schema.ClientMeta, parent *schema.Resource, res chan<- interface{}) error {
 	var config iam.ListAccessKeysInput
-	p := parent.Item.(UserWrapper)
+	p := parent.Item.(*types.User)
 	svc := meta.(*client.Client).Services().IAM
 	if aws.ToString(p.UserName) == rootName {
 		return nil
```
Comment (bbernays) [inline plugins/source/aws/resources/services/iam/users_fetch.go:104]: Not necessary anymore as Root user is never returned

File: plugins/source/aws/resources/services/iam/user_policies_fetch.go [hunk 5]
```diff
@@ -14,7 +15,7 @@ import (
 func fetchIamUserPolicies(ctx context.Context, meta schema.ClientMeta, parent *schema.Resource, res chan<- interface{}) error {
 	c := meta.(*client.Client)
 	svc := c.Services().IAM
-	user := parent.Item.(UserWrapper)
+	user := parent.Item.(*types.User)
 	if aws.ToString(user.UserName) == rootName {
 		return nil
 	}
```
Comment (bbernays) [inline plugins/source/aws/resources/services/iam/user_policies_fetch.go:21]: Not necessary

File: plugins/source/aws/resources/plugin/tables.go [hunk 6]
```diff
@@ -195,6 +195,7 @@ func tables() []*schema.Table {
 		glue.Workflows(),
 		guardduty.Detectors(),
 		iam.Accounts(),
+		iam.CredentialReportUsers(),
```
Comment (bbernays) [inline plugins/source/aws/resources/plugin/tables.go:198]: I would call this just `iam.CredentialReport`

File: plugins/source/aws/resources/services/iam/credential_report_users_fetch.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	PasswordNextRotation  string    `csv:"password_next_rotation"`
+	MfaActive             bool      `csv:"mfa_active"`
+	AccessKey1Active      bool      `csv:"access_key_1_active"`
+	AccessKey2Active      bool      `csv:"access_key_2_active"`
+	AccessKey1LastRotated string    `csv:"access_key_1_last_rotated"`
+	AccessKey2LastRotated string    `csv:"access_key_2_last_rotated"`
+	Cert1Active           bool      `csv:"cert_1_active"`
+	Cert2Active           bool      `csv:"cert_2_active"`
+	Cert1LastRotated      string    `csv:"cert_1_last_rotated"`
+	Cert2LastRotated      string    `csv:"cert_2_last_rotated"`
```
Comment (bbernays) [inline plugins/source/aws/resources/services/iam/credential_report_users_fetch.go:29]: There are some fields that are missing:

```
access_key_1_last_used_date
access_key_1_last_used_region
access_key_1_last_used_service
access_key_2_last_used_date
access_key_2_last_used_region
access_key_2_last_used_service
password_last_used
```

File: plugins/source/aws/resources/plugin/tables.go (hunk 6 above)
Comment (hermanschaaf) [reply plugins/source/aws/resources/plugin/tables.go:198]: Hmm, I'm just worried that the table will then be called `credential_reports`, which doesn't sound right considering that it contains user entries, not report entries. WDYT?

File: plugins/source/aws/resources/services/iam/credential_report_users_fetch.go (hunk 7 above)
Comment (hermanschaaf) [reply plugins/source/aws/resources/services/iam/credential_report_users_fetch.go]: Added 👍

File: plugins/source/aws/resources/plugin/tables.go (hunk 6 above)
Comment (bbernays) [reply plugins/source/aws/resources/plugin/tables.go:198]: Users are the only resource that have credentials though

File: plugins/source/aws/resources/services/iam/credential_reports.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+			},
+			{
+				Name:     "user",
+				Type:     schema.TypeString,
+				Resolver: schema.PathResolver("User"),
+			},
+			{
+				Name:     "user_creation_time",
+				Type:     schema.TypeTimestamp,
+				Resolver: schema.PathResolver("UserCreationTime"),
```
[the remark this answers is not in the document]
Comment (hermanschaaf) [reply plugins/source/aws/resources/services/iam/credential_reports.go:32]: Good catch, updated!

**Review by bbernays — APPROVED** [review]
LGTM

**Pull request merged**